### ISSUE TITLE
Scratch buffers

### ISSUE THREAD

**Issue by rwols** [microsoft/language-server-protocol#1030]
There's no schema for scratch buffers (a buffer in the editor that has no associated file).

I propose something like `buffer://<some-string-id-provider-by-client>` to tell a language server: "Here's a document that's opened, but it has no associated file on disk".

Because the content of the buffer is provided in the textDocument/didOpen notification there shouldn't be a technical roadblock to this.

Because a server knows the workspace folders, it can deduce some basic properties like find an associated config file (e.g. tsconfig.js). So it should be able to provide basic services to scratch buffers.

**Comment by rwols** [comment]
Once the buffer is saved to disk by the client, the client should send a textDocument/didClose with the text document identifier being `buffer://...`, followed by a textDocument/didOpen with the text document item of the now saved-to-disk file.

**Comment by floitsch** [comment]
Isn't that the same as the `untitled:Untitled-1` that is used by vscode?

**Comment by dbaeumer** [comment]
Yes, but it is not as a standard in the spec so it might depend on how every client implements this.

**Comment by kaddkaka** [comment]
Any chance this might get into the specification?

**Comment by dbaeumer** [comment]
A PR is welcome but we need an implementation in a server and client as well before we add things to the spec.
